Enable userns by default

cpuguy83 opened the pull request:
Posting this for discussion.
Note that LXC already defaults to an equivalent setting. It would be nice to push things along here.

This is a huge boost to default security.
Some things to work out:

- [x] Option to allow auto userns=host for privileged and net/pid/ipc=host
- [x] Support mapping uid/gids for container mounts to userns
- [x] Make mount id mapping optional (disabled by default?)
- [x] Make mount mapper pluggable
- [x] Support non-recursive bind mounts for idmapfs
- [x] Unmount mapped fs on container shutdown
- [ ] Checkpoint/restore (currently broken)
- [x] Address with userns=host container fs has ownership set to the remapped root instead of real root.
- [ ] Fix issue with docker-py tests relying on vieux/sshfs which is broken on userns
- [x] With a mapping fs, there is no need to chown image layers when they are loaded
- [ ] Handle unmounting mapped mounts (on exit, on failure cases, etc)
- [ ] Backwards compatibility
- [ ] Figure out what and how to expose various options

cpuguy83 committed:
Enable userns by default

This is a huge boost to default seucrity.
Some things to work out:

- [ ] How to not break users who already have unmapped images loaded
- [ ] How to not break users who mount stuff from the host
- [ ] Should `--privileged` mean `--userns=host` as well to prevent
breakage?

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Also another benefit of this discussion, at the very least some of this work can help make running with remapped root enabled easier than it currently is regardless of the default.

thaJeztah commented:
One blocker may be CentOS / RHEL, where I think it's not supported with the default configuration; not sure if this complicates running k8s (I seem to recall it uses a lot of bind-mounts which may be problematic without shiftfs)

cpuguy83 commented:
@thaJeztah 

RHEL/CentOS: We can check the setting (it's in /proc/sys/kernel somewhere if I remember correctly) and handle/warn accordingly (`[WARN] defaulting to insecure/no user namespace support due to kernel support, enable support by setting /proc/sys/kernel/... and restart dockerd`)

k8s: I think there is likely to be issues for many people (not just k8s), which we can try to mitigate through various things (fuse), but ultimately this is why the flag to disable is there. We just need to make more noise on whatever release this default changes so people can make the required configurations before deploying new systems with the new version.

tonistiigi commented:
fyi https://github.com/moby/buildkit/issues/766

cpuguy83 committed:
Use idmapfs to map host bind mounts

idmapfs allows us to create a bind mount with the ownership of the
file/tree mapped to the container's user namespace mapping.

This change makes use of idmapfs to automatically map user requested
host bind mounts to the container's ID space.

For (a really bad) example, with this change with user namespces enabled
you can now do `docker run --rm --user=root -v /etc:/hostetc busybox cat
/hostetc/shadow`

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Looks like there's some issues with the plugin tests with userns.
Seems the last time CI ran for userns was *ahem* ~1.5yr ago. /cc @andrewhsu

---

I've also added a 2nd commit for testing purposes... it does uid/gid mapping for bind mounts into the container's ID space using https://github.com/cpuguy83/idmapfs
Right now it does all user specified mounts, even if it may not be necessary to do so.

cpuguy83 commented:
Seems breakage is likely caused by 28334c1d829b9ea8ac7e10d45a6db37c3821d5db

cpuguy83 committed:
Fix test daemon config for userns path

When starting up the test daemon we set data root to some random dir,
and then update that dir after the daemon is started to whatever is
reported from calling the info API.
When the daemon is restarted we then use that updated path as the data
root for the new daemon process.

In the case of user namespaces, the daemon actually repoprts a new root
directory... e.g.

With this:

```
dockerd --data-root=/foo/bar --userns-remap=default
```

The daemon root gets reported as `/foo/bar/<mapped uid>.<mapped gid>`.
This means when we restart the test daemon we set a new data root of
`/foo/bar/<maped uid>.<mapped gid>` which then the daemon will update
again to `/foo/bar/<maped uid>.<mapped gid>/<maped uid>.<mapped gid>`
which causes tests to fail expecting state to stick around.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Fix plugin graphdriver tests for userns

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Add daemon opts to set auto userns=host certain conditions

These options allow an admin to preserve compatability with old behavior
when userns is enabled.
THey are disabled by default for security purposes.

Specifically this adds daemon options to automatically set userns=host
when:

1. --net=host
2. --pid=host
3. --privileged

The config change is not actually set on the container config so the
options can be turned off and have the policy change enforced without
recreating the container.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Fix domain and C/R tests to use userns=host

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Fix userns remapped file mounts

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

AkihiroSuda commented:
>  Option to allow auto userns=host for privileged and net/pid/ipc=host

and for containers with mounts, until we can get in-kernel idmapfs?

cpuguy83 commented:
It's on the todo checklist. 👼❤️

cpuguy83 committed:
Add option to enable userns filesystem remapping

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Looks like docker-py tests are using viex/sshfs which is mounting /var/lib/docker/plugins, which doesn't exist for userns (because it's under /var/lib/docker/<remappedroot>/plugins).
It's not clear to me why it needs it at all... but probably this could use a different plugin anyway.

cpuguy83 committed:
fix WithNamespaces after rebase

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Change DisableUsernsRemap to just set the mode

Currently the only mode is "host", and "anything that's not host", where
"host" is basically the old default.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Kinda cool, updated overlay2 to support re-mapping using idmapfs instead of chown.

cpuguy83 commented:
Also started working on a driver abstraction so that we an throw in other mapping drivers (e.g. shifts on ubuntu's kernel), however the current way mapping and generally all filesystem handling makes this pretty difficult to actually implement.

cpuguy83 committed:
[WIP] Map container rootfs using mapping driver

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Handling mapping the rootfs more generically now... any graphdriver driver will be automatically mapped. This also fixes issues with `--userns=host` having the wrong file ownership (this is a problem with the chowning approach before this change).

I think I can now actually handle the mapped volume/bind mounts just under a single mount from the container's rootfs instead of special casing them... wondering if I should do this or not. I currently expose an option to enable auto-mapping of volumes/binds to the container's user namespace... mapping the resolved rootfs would make mapping  non-optional (currently only optional at the daemon level).

cpuguy83 commented:
Would appreciate some input here from others.
From a usability perspective, automapping seems natural, but I'm not sure if some people would want to *not* do mapping.

cpuguy83 commented:
So perhaps something useful to extract from this into it's own PR: Support for setting filesystem mappings for mounts... e.g. `--mount type=bind,source=/foo,target=/bar,uidmap=...,gidmap=...`

Likely uidmap and gidmap would need to support multiple entries.
I think this is useful outside of user namespaces so people don't have to try and match uid's for all containers.
The UI I mentioned above isn't *great*.

jleadford commented:
@cpuguy83, could you explain the motivation behind using https://github.com/cpuguy83/idmapfs
for remapping w/r/t mounts, versus a chown-based approach?

The chown-based approach is used by e.g. LXC
https://github.com/lxc/lxc/blob/d6a4505542bb5702c293033ee1f5b6dc842420bc/src/lxc/storage/overlay.c#L161

and I'd contend it's well-understood.  

I admit that I'm not familiar with your idmapfs approach (will endeavor review the code), but given the caveats listed at https://github.com/cpuguy83/idmapfs#status (new project, 'should not be considered secure or stable'), I'm weary of relying on it for something that'll be docker-default. 

There's also use of POSIX ACLs with setfacl, e.g. https://github.com/lxc/lxd/issues/714#issuecomment-110934439, which could be explored for some cases.

cpuguy83 commented:
@jleadford 

idmapfs is there as a placeholder, potentially a fallback. We need to be able to support mapping host paths and volumes into the user namespace and have the ownership line up. Ideally this would be something like shiftfs.
This is not something we can do with `chown`.

`chown` for images kind of works today, but:

1. Poses an issue for running a container with the host userns
2. Has significant storage impact in the case of containers each having their own id mappings (not currently supported today, but is at least a long-term goal).

jleadford commented:
Just so I'm clear here, why isn't chown (or POSIX ACLs to extend access to the right IDs) possible on host paths?

I know it's not a natural approach (compared to, let's say, kernel support for this sort of thing), just wondering though.

jleadford commented:
The forthcoming Linux Plumbers Conference in Lisbon will have a session on shiftfs https://linuxplumbersconf.org/event/4/sessions/57/#20190910, which is what LXD is now using for mount-remapping https://discuss.linuxcontainers.org/t/seamless-filesystem-uid-gid-shift-and-invert-shift/3369, https://discuss.linuxcontainers.org/t/lxd-3-16-has-been-released/5445. 

The session seeks to discuss plans for this feature generally in mainline, and which impl to go with, see e.g. https://twitter.com/stgraber/status/1160214145918210048. The output of this session is pertinent to this effort.

I agree wtih @AkihiroSuda, though. I don't think this (lack of current support in mainline) holds this PR back; for the meantime, containers with mounts can have userns=host set. I think this is a pragmatic choice for now.

jleadford commented:
hey @cpuguy83 , thoughts on the above?

cpuguy83 commented:
@jleadford Apologies for the delayed response, I've been on vacation.

re: modifying host paths
Changing the host paths like this is generally undesirable as it can have unexpected consequences for the user and even conflicting modifications between two different container specs.

Example of this are the  ":z" and ":Z" flags that we support for bind mounted host paths, which sets selinux labels on the host paths.

AkihiroSuda commented:
Can we also support shiftfs? (only merged in Ubuntu kernel though)

omeid commented:
@AkihiroSuda I think that can happen as a separate PR. This is a large enough chunk of work on it's own.

AkihiroSuda commented:
:+1:

Maybe this PR can be also decompose into a set of small PRs before making it default.
Especially `Address with userns=host container fs has ownership set to the remapped root instead of real root.` .

AkihiroSuda commented:
> Maybe this PR can be also decompose into a set of small PRs before making it default.
> Especially Address with userns=host container fs has ownership set to the remapped root instead of real root. .

@cpuguy83 WDYT?

cpuguy83 commented:
@AkihiroSuda Yes indeed, making it default is a specific decision that should be separate from making it simpler.
Making it default in as this is being tested, though, is the only way to get all the CI to run with userns enabled to figure out what is broken still.

If anyone is looking to move this forward faster, I could use some help.
These are really stressful times and I've found that I am unable to juggle as many things as before.
